### PR TITLE
Add a module for ZFS pool state

### LINKED ISSUE #5590: zfs pool (harddrive) notifications
Hello, I'm not a good programmer and would like to ask if someone could bring up a feature which alert me when a pool is failed or degraded 'cause a hard drive riches end of life? 

best Regards, Dan

### PR DESCRIPTION
**Pull request by vlvkobal**
##### Summary

SSIA

Fixes #5590

##### Component Name
proc plugin

##### Test Plan
Create a ZFS pool. Check the `zfspool.state_<pool_name>` chart.
Try to create/delete ZFS pools and check if the charts are correctly created/obsoleted.

### CODE REVIEW

**Review by thiagoftsm — APPROVED** [review]

File: collectors/proc.plugin/proc_spl_kstat_zfs.c [hunk 1]
```diff
@@ -194,3 +198,206 @@ int do_proc_spl_kstat_zfs_arcstats(int update_every, usec_t dt) {
...
+                        pool->faulted = 1;
+                    } else if (!strcmp(state, "OFFLINE\n")) {
+                        pool->offline = 1;
+                    } else if (!strcmp(state, "REMOVED\n")) {
+                        pool->removed = 1;
+                    } else if (!strcmp(state, "UNAVAIL\n")) {
+                        pool->unavail = 1;
+                    } else {
+                        disable_zfs_pool_state(pool);
+                        error("ZFS POOLS: Undefined state for zpool %s, disabling the chart", de->d_name);
```
Comment (ilyam8) [inline collectors/proc.plugin/proc_spl_kstat_zfs.c:375]: let's log the `state` value

Comment (vlvkobal) [reply collectors/proc.plugin/proc_spl_kstat_zfs.c]: Done.

**Review by thiagoftsm — DISMISSED** [review]

File: collectors/proc.plugin/proc_spl_kstat_zfs.c [hunk 2]
```diff
@@ -194,3 +198,206 @@ int do_proc_spl_kstat_zfs_arcstats(int update_every, usec_t dt) {
...
+                        pool->faulted = 1;
+                    } else if (!strcmp(state, "OFFLINE\n")) {
+                        pool->offline = 1;
+                    } else if (!strcmp(state, "REMOVED\n")) {
+                        pool->removed = 1;
+                    } else if (!strcmp(state, "UNAVAIL\n")) {
+                        pool->unavail = 1;
+                    } else {
+                        disable_zfs_pool_state(pool);
+                        error("ZFS POOLS: Undefined state %s for zpool %s, disabling the chart", state, de->d_name);
```
Comment (ilyam8) [inline collectors/proc.plugin/proc_spl_kstat_zfs.c:375]: The `state` value contains newline character (\n). I think we want to remove it when logging an undefined state?

File: web/gui/dashboard_info.js [hunk 3]
```diff
@@ -1490,6 +1496,13 @@ netdataDashboard.context = {
         info: 'inodes (or index nodes) are filesystem objects (e.g. files and directories). On many types of file system implementations, the maximum number of inodes is fixed at filesystem creation, limiting the maximum number of files the filesystem can hold. It is possible for a device to run out of inodes. When this happens, new files cannot be created on the device, even though there may be free space available.'
     },
 
+    // ------------------------------------------------------------------------
+    // ZFS pools
+    'zfspool.state': {
+        info: 'ZFS pool state. The overall health of a pool, as reported by <code>zpool status</code>, is determined by the aggregate state of all devices within the pool. ' +
+            'For details, see <a href="https://docs.oracle.com/cd/E19253-01/819-5461/gamno/index.html" target="_blank"> ZFS documentation</a>.'
```
Comment (ilyam8) [inline web/gui/dashboard_info.js:1503]:
```suggestion
            'For details, see <a href="https://openzfs.github.io/openzfs-docs/man/8/zpoolconcepts.8.html?#Device_Failure_and_Recovery" target="_blank"> ZFS documentation</a>.'
```

See https://github.com/netdata/netdata/issues/5590#issuecomment-831792101

**Comment by ilyam8** [comment]
The code looks good, to make the PR even better we could:
 - [docs] add ZFS pools to [the proc plugin readme](https://github.com/netdata/netdata/tree/master/collectors/proc.plugin#procplugin).
 - [alarms] see https://github.com/netdata/netdata/issues/5590#issuecomment-831769138
   - `warn` when DEGRADED
   - `crit` when FAULTED or UNAVAIL.

File: collectors/proc.plugin/proc_spl_kstat_zfs.c [hunk 4]
```diff
@@ -194,3 +198,206 @@ int do_proc_spl_kstat_zfs_arcstats(int update_every, usec_t dt) {
...
+                snprintfz(
+                    filename, FILENAME_MAX, "%s%s/%s/state", netdata_configured_host_prefix, ZFS_PROC_POOLS, de->d_name);
+
+                char state[STATE_SIZE + 1];
+                int ret = read_file(filename, state, STATE_SIZE);
+
+                if (!ret) {
+                    state_file_found = 1;
+
+                    // ZFS pool states are described at https://docs.oracle.com/cd/E19253-01/819-5461/gamno/index.html
```
Comment (ilyam8) [inline collectors/proc.plugin/proc_spl_kstat_zfs.c:360]:
```suggestion
                    // ZFS pool states are described at https://openzfs.github.io/openzfs-docs/man/8/zpoolconcepts.8.html?#Device_Failure_and_Recovery
```

using official OpenZFS documentation sounds better

File: collectors/proc.plugin/proc_spl_kstat_zfs.c (hunk 2 above)
Comment (vlvkobal) [reply collectors/proc.plugin/proc_spl_kstat_zfs.c]: Fixed.

File: collectors/proc.plugin/proc_spl_kstat_zfs.c [hunk 5]
```diff
@@ -372,6 +372,10 @@ int do_proc_spl_kstat_zfs_pool_state(int update_every, usec_t dt)
                         pool->unavail = 1;
                     } else {
                         disable_zfs_pool_state(pool);
+
+                        char *c = strchr(state, '\n');
```
Comment (ilyam8) [inline collectors/proc.plugin/proc_spl_kstat_zfs.c:378]: neat: how about doing it before we if/elseif ?

Comment (vlvkobal) [reply collectors/proc.plugin/proc_spl_kstat_zfs.c:378]: I don't like the idea. Actually, `strchr` is a loop. I don't see the point of doing it on every file read.

**Review by thiagoftsm — APPROVED** [review]

**Review by ilyam8 — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]

**Pull request merged**